crates/ripemd160/src/ripemd160.rs

```diff
@@ -0,0 +1,408 @@
+            cs.namespace(|| format!("first add_many in compute_f: round {i}, left = {left}",)),
+            &[
+                msg_digest[0].clone(),
+                f.clone(),
+                msg_words[msg_word_index].clone(),
+                UInt32::constant(round_constant),
+            ],
+        )?;
+
+        // t = rotate_left(t, rotl_amount)
+        t = uint32_rotl(&t, rotl_amount);
+
+        // t = t + E
+        t = UInt32::addmany(
+            cs.namespace(|| format!("second add_many in compute_f: round {i}, left = {left}",)),
+            &[t, msg_digest[4].clone()],
+        )?;
+
+        // next_msg_digest = [E, t, B, rotate_left(C, 10), D]
+        msg_digest[0] = msg_digest[4].clone();
+        msg_digest[4] = msg_digest[3].clone();
+        msg_digest[3] = uint32_rotl(&msg_digest[2], 10);
+        msg_digest[2] = msg_digest[1].clone();
+        msg_digest[1] = t;
+    }
+    Ok(())
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use bellpepper::gadgets::multipack::bytes_to_bits;
+    use bellpepper_core::{boolean::AllocatedBit, test_cs::TestConstraintSystem};
+    use hex_literal::hex;
+    use pasta_curves::Fp;
+    use rand_core::{RngCore, SeedableRng};
+    use rand_xorshift::XorShiftRng;
+
+    #[test]
+    fn test_blank_hash() {
+        let msg = "".as_bytes();
+        let msg_bits = bytes_to_bits(msg);
+
+        let mut cs = TestConstraintSystem::<Fp>::new();
+        let input_bits: Vec<Boolean> = msg_bits
+            .into_iter()
+            .map(Boolean::Constant)
+            .collect::<Vec<_>>();
+
+        let out_bits = ripemd160(cs.namespace(|| "ripemd160"), &input_bits).unwrap();
+
+        assert!(cs.is_satisfied());
+        assert_eq!(cs.num_constraints(), 0);
+
+        let expected = hex!("9c1185a5c5e9fc54612808977ee8f548b2258d31");
+        let mut out = out_bits.iter();
+        for b in expected.iter() {
+            for i in 0..8 {
+                let c = out.next().unwrap().get_value().unwrap();
+                assert_eq!(c, (b >> (7 - i)) & 1u8 == 1u8);
+            }
+        }
+    }
+
+    #[test]
+    fn test_hash_abc() {
+        let msg = "abc".as_bytes();
+        let msg_bits = bytes_to_bits(msg);
+
+        let mut cs = TestConstraintSystem::<Fp>::new();
+        let input_bits: Vec<Boolean> = msg_bits
+            .into_iter()
+            .map(Boolean::Constant)
+            .collect::<Vec<_>>();
+
+        let out_bits = ripemd160(cs.namespace(|| "ripemd160"), &input_bits).unwrap();
+
+        assert!(cs.is_satisfied());
+        assert_eq!(cs.num_constraints(), 0);
+
+        assert!(cs.is_satisfied());
+        let expected = hex!("8eb208f7e05d987a9b044a8e98c6b087f15a0bfc");
+        let mut out = out_bits.iter();
+        for b in expected.iter() {
+            for i in 0..8 {
+                let c = out.next().unwrap().get_value().unwrap();
+                assert_eq!(c, (b >> (7 - i)) & 1u8 == 1u8);
+            }
+        }
+    }
```

Comment:
It would be great to see a randomized test here using an out-of-circuit implementation of ripemd160, and checking your gadget returns the same. You have the scaffolding for randomized tests already, and you can use the https://github.com/RustCrypto/hashes/tree/master/ripemd crate in rust to compute a reference value.

See e.g. https://github.com/lurk-lab/bellpepper-gadgets/blob/main/crates/keccak/tests/gadget.rs#L51-L83 for this sort of test (using different tooling for the RNG portion, i.e. proptest - though what you have here is fine, the important part is comparison to a reference implementation).